components/scream/src/share/field/field.hpp

```diff
@@ -200,45 +263,89 @@ template<typename RealType>
-  constexpr int DstRank = ekat::GetRanks<DT>::rank;
+  // Start by reshaping into a view with all dyn extents
+  const auto view_ND = get_ND_view<DstValueType,DstRank>();
 
-  // Check the reinterpret cast makes sense for the two value types (need integer sizes ratio)
-  EKAT_REQUIRE_MSG(alloc_prop.template is_allocation_compatible_with_value_type<DstValueType>(),
-                       "Error! Source field allocation is not compatible with the destination field's value type.\n");
+  constexpr int DstRankDynamic= DstView::rank_dynamic;
 
-  // The destination view type
-  using DstView = ekat::Unmanaged<typename KokkosTypes<DefaultDevice>::template view<DT> >;
-  typename DstView::traits::array_layout kokkos_layout;
-
-  const int num_values = alloc_prop.get_alloc_size() / sizeof(DstValueType);
-  if (DstRank==1) {
-    // We are staying 1d, possibly changing the data type
-    kokkos_layout.dimension[0] = num_values;
-  } else {
-    int num_last_dim_values = num_values;
-    // The destination data type is a multi-dimensional array.
-    for (int i=0; i<field_layout.rank()-1; ++i) {
-      kokkos_layout.dimension[i] = field_layout.dim(i);
-
-      // Safety check: field_layout.dim(0)*...*field_layout.dim(field_layout.rank()-2) should divide num_values, so we check
-      EKAT_REQUIRE_MSG(num_last_dim_values % field_layout.dim(i) == 0, "Error! Something is wrong with the allocation properties.\n");
-      num_last_dim_values /= field_layout.dim(i);
+  using dyn_DT = typename decltype(view_ND)::traits::data_type;
+  if (!std::is_same<dyn_DT,DT>::value) {
+    // The user requested some compile-time dimensions.
+    // Let's check that they are correct
+    for (int i=DstRankDynamic; i<DstRank; ++i) {
+      EKAT_REQUIRE_MSG(view_ND.extent(i)==dims_type::static_extent(i),
+          "Error! The template DataType contains some invalid compile-time dimensions.\n");
     }
-    kokkos_layout.dimension[field_layout.rank()-1] = num_last_dim_values;
   }
 
-  return DstView (reinterpret_cast<DstValueType*>(m_view.data()),kokkos_layout);
+  // Before building the DstView from view_ND, we have one more check:
+  // if DstRankDynamic==0, kokkos specializes the view offset struct,
+  // assuming *no* stride. That's fine, as long as this field alloc
+  // props ensure that there is no stride
+  EKAT_REQUIRE_MSG (DstRankDynamic>0 || alloc_prop.contiguous(),
+      "Error! Cannot use all compile-time dimensions for strided views.\n");
+
+  return DstView(view_ND);
+}
+
+template<typename RealType>
+Field<RealType> Field<RealType>::
+subfield (const std::string& sf_name, const int idim, const int k) const {
+  const auto& id = m_header->get_identifier();
+  const auto& lt = id.get_layout();
+
+  // Sanity checks
+  EKAT_REQUIRE_MSG (is_allocated(),
+      "Error! Input field must be allocated in order to subview it.\n");
+  EKAT_REQUIRE_MSG (idim==0 || idim==1,
+        "Error! Subview dimension index must be either 0 or 1.\n");
```

Comment:
This is to simplify the implementation in terms of assumptions made to provide padding for layouts, right? Not that it will surprise anyone, but we should probably document this limitation above the declaration of this method.

Reply:
You mean the idim=0,1 constraint? The reason is that if you subview at idim=2+, then there is no way to keep LayoutRight. In Kokkos, LayoutRight allows to store a single stride for the slowest dim, without hurting performance at all (LR computes offsets like `offset = i7 + dim(6)*(i6+dim(5)* (...)) + stride*i0`, where `stride=dim(1)*...*dim(7)` for contiguous views).

If you slice with idim=2, you can still fit in this offset strategy, but if you slice with idim=2+, you can't, and need LayoutStride. Besides the perf loss, I don't think we'd be able to implement it, since the return type would be different.

I can document better in the Field class, where the function is declared. Better yet, make the fcn private, and use a friend free function in field_utils.hpp, with ample documentation (the file field.hpp is too crowded).